protos/protos/network.proto

```diff
@@ -0,0 +1,46 @@
+syntax = "proto3";
+
+import "protos/chain.proto";
+
+message PeerInfo {
+    bytes id = 1;
+    oneof addr {
```

Comment:
why are you using oneof { bool, some } for optional? you could just say addr.. optional is implied in proto3

Reply:
Yeah I realize that at some point but forgot to change it :)

Reply:
I guess one advantage if that you don't have to check whether the field is set. But probably not worth it.

Reply:
yea. i don't think you wanna do that. it also adds bytes to the proto header which has to get sent with every message

Reply:
It seems that the generated proto class does not provide API to check whether a field is set, if the field has a primitive type.

Reply:
are you concerned about someone writing code that omits setting a value and this going unnoticed?

Reply:
No. I am concerned with how you check whether a field is `None`.

Reply:
i see.. it looks like you may have to explicitly specify optional for rust-protobuf to specify the field as optional in the struct. can you try this? i don't remember if the optional keyword is compatible with protobuf 3.

Reply:
https://stackoverflow.com/questions/42622015/how-to-define-an-optional-field-in-protobuf-3 It seems that there is no way to do this

Reply:
Let's not use protobuf 2, unless we are using it everywhere.
By default in protobuf 3 - everything is optional, and that optional is just represented by default value of the type. In the case of fields here - empty string would mean `None`.
@azban proto3 removed `has_field` for primitive types. Actual recommendation for `optional` primitive types is to use structs which do have `has_field`:
`
Optional isn't removed; everything is optional in proto3. But yes, field visibility (has_field) has been removed for primitives. If you need field visibility use wrappers.proto which has messages like StringValue. Since they are messages, has_field is available. This is effectively "boxing" which is common in many languages.
`
from https://stackoverflow.com/questions/31801257/why-required-and-optional-is-removed-in-protocol-buffers-3

Reply:
doing null checks by value is not rust-like .. i don't really see an issue with using proto2 for network stuff until we want to figure out how to make rust proto generation work with proto3 as intended.

Reply:
Again, nothing wrong with rust proto.
Proto3 specification removed the check for primitive types. See comment above.

Reply:
rust proto should make every field an `Option` field.. checking based on some default for each value type is not rust-like

Reply:
Changed to use wrapper types. Rust protobuf actually has support for that https://docs.rs/protobuf/2.4.0/protobuf/well_known_types/index.html